File: MainBackend/Services/BillService.cs
using MainBackend.Database.Entities;
using MainBackend.Database.MainDb.UoW;
using MainBackend.DTO;
using MainBackend.Enums;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using MainBackend.Exceptions;
using Microsoft.AspNetCore.Http.HttpResults;

namespace MainBackend.Services;

public class BillService: IBillService
{
    private readonly IIdentityService identityService;
    private readonly IUoW uow;

    public BillService(IUoW uow, IIdentityService identityService)
    {
        this.identityService = identityService;
        this.uow = uow;
    }

    public async Task<List<KeyValuePair<string, string>>> GetAllCurrencies()
    {
        return await Task.FromResult(CurrencyExtensions.GetAllCurrencies());
    }
    
    public async Task<BillResponse> CreateBill(BillDTO billDto, int groupId)
    {
        var newBill = new Bill(billDto);
        uow.BillRepository.Create(newBill);
        
        var group = await uow.GroupRepository.Get(groupId);
        if (group == null)
        {
            throw new Exception("Group does not exist");
        }
        
        group.Bills.Add(newBill);
        uow.GroupRepository.Update(group);
        await uow.Save();
        
        return new BillResponse { BillId = newBill.Id };
    }

    public async Task AddMenuItemsToBillAsync(int billId, List<MenuItem> menuItems)
    {
        if (menuItems == null || !menuItems.Any())
        {
            throw new ArgumentException("No menu items provided.");
        }

        var bill = await uow.BillRepository.GetBillByIdAsync(billId);
        if (bill == null)
        {
            throw new KeyNotFoundException($"Bill with ID {billId} not found.");
        }

        await uow.BillRepository.AddMenuItemsToBillAsync(bill, menuItems);
    }


    public async Task UpadateNameBill(string billName, int billId)
    {
        var bill = await uow.BillRepository.GetBillByIdAsync(billId);
        bill.Name = billName;
        uow.BillRepository.Update(bill);
        await uow.Save();
    }

    public async Task UpadateDateBill(DateTime date, int billId)
    {
        var bill = await uow.BillRepository.GetBillByIdAsync(billId);
        bill.Date = date;
        uow.BillRepository.Update(bill);
        await uow.Save();
}
    public async Task<BillDetailDTO> GetBillDetailsAsync(int billId)
    {
        try
        {
            var bill = await uow.BillRepository.GetBillDetails(billId);
            if (bill == null)
            {
                return null;
            }

            // Calculate total amount from menu items
            decimal totalAmount = bill.MenuItems?.Sum(item => item.Price * (item.Quantity ?? 1)) ?? 0;

            // Get the payer information
            string paidBy = GetPaidByInfo(bill);

            var menuItems = MapMenuItems(bill.MenuItems);

            return new BillDetailDTO
            {
                Id = bill.Id,
                Name = bill.Name ?? "Unnamed bill",
                Image = bill.BillImage ?? "default-receipt.png",
                Date = bill.Date,
                Amount = totalAmount,
                Status = "pending", // bill.Status.ToString().ToLower(),
                Currency = bill.Currency.ToString(),
                PaidBy = paidBy,
                Items = menuItems
            };
        }
        catch (Exception ex)
        {
            throw new Exception($"Error retrieving bill details: {ex.Message}", ex);
        }
    }

    private string GetPaidByInfo(Bill bill)
    {
        if (bill.Payments?.Any() != true || bill.Payments.First()?.User == null)
        {
            return "Not assigned";
        }
        return bill.Payments.First().User.Name;
    }

    private List<MenuItemDetailDTO> MapMenuItems(ICollection<MenuItem> menuItems)
    {
        if (menuItems == null)
        {
            return new List<MenuItemDetailDTO>();
        }

        return menuItems.Select(item => new MenuItemDetailDTO
        {
            Id = item.Id,
            Name = item.Name ?? "Unnamed item",
            Price = item.Price,
            Quantity = item.Quantity ?? 1,
            AssignedTo = item.OrderedBy?.Select(user => new UserDTO
            {
                Id = user.Id,
                Name = user.Name,
                LastName = user.LastName,
                Email = user.EmailAddress,
                Image = user.Image ?? "default-avatar.png"
            }).ToList() ?? new List<UserDTO>()
        }).ToList();

    }

    public async Task AddPaymentToBill(int billId, int userId)
    {
        var bill = await uow.BillRepository.GetBillByIdAsync(billId);
        if (bill == null)
        {
            throw new KeyNotFoundException("Bill not found.");
        }
        
        
        var user = await uow.UserRepository.GetByIdAsync(userId);
        if (user == null)
        {
            throw new KeyNotFoundException("User not found.");
        }
        decimal totalAmount = bill.MenuItems?.Sum(item => item.Price * (item.Quantity ?? 1)) ?? 0;
        var payment = new Payment
        {
            User = user,
            UserId = user.Id,
            Amount = totalAmount,
        };
        
        bill.Payments ??= new List<Payment>();
        bill.Payments.Add(payment);

        
         uow.BillRepository.Update(bill);
         await uow.Save();
         
    }

    public async Task<decimal> GetMySumInBill(int billId)
    {
        var bill = await uow.BillRepository.GetBillWithOredrByByIdAsync(billId);
        if (bill == null)
        {
            throw new KeyNotFoundException("Bill not found.");
        }
        var userId = identityService.GetLoggedUserId();
        
        decimal totalSum = (decimal)(bill.MenuItems?
            .Where(item => item.OrderedBy.Any(user => user.Id == userId))
            .Sum(item => item.Price * (item.Quantity ?? 1)) ?? 0);

        return totalSum;
    }

    public async Task UpdateBillImage(string billImage, int billId)
    {
        Bill bill=await uow.BillRepository.Get(billId);
        if (bill == null)
            throw new NotFoundException();
        bill.BillImage= billImage;
        uow.BillRepository.Update(bill);
        await uow.Save();
    }

}

File: frontendweb/src/services/billService.ts
import { Bill, BillStatus, MenuItem, User } from '@/types';
import { api } from './api';
import internal from 'stream';

export interface CreateBillRequest {
  name: string;
  date: string;
  amount: number;
  currency: string;
  paidBy: string;
  status: BillStatus;
  image?: string;
}
export interface BillIdObject {
  billId: number;
}
export interface BillResponse {
  billId: BillIdObject;
  menuItems: MenuItem[];
}

export interface BillDetailsResponse {
  id: number;
  name: string;
  image: string;
  date: string;
  amount: number;
  status: string;
  currency: string;
  paidBy: string; 
  items: {
    id: number;
    name: string;
    price: number;
    quantity: number;
    assignedTo: User[];
  }[];
}

export const billService = {
  getBill: async (id: number): Promise<BillDetailsResponse> => {
    const response = await api.get(`/bill/${id}`);
    return response.data;
  },

  async getAllCurrencies(): Promise<{ key: string; value: string }[]> {
    const response = await api.get<{ key: string; value: string }[]>('/bill');
    return response.data;
  },
  createBill: async (groupId: number, formData: FormData): Promise<{ billId: BillIdObject; menuItems: MenuItem[] }> => {
    const response = await api.post<{ billId: BillIdObject; menuItems: MenuItem[] }>(`/bill/${groupId}`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
    return response.data;
  },
  // async createBill(groupId: number, bill: CreateBillRequest, image: File): Promise<BillResponse> {
  //   const formData = new FormData();
  //   formData.append('image', image);
  //   Object.entries(bill).forEach(([key, value]) => {
  //     formData.append(key, value.toString());
  //   });

  //   const response = await api.post<BillResponse>(`/bill/${groupId}`, formData);
  //   return response.data;
  // },
  async addMenuItemsToBill(billId: number, menuItems: MenuItem[]) {
    const payload = menuItems.map(item => ({
      id: item.id,
      name: item.name,
      price: item.price,
      quantity: item.quantity,
      orderedBy: [],
      transfers: [],
      status: 0,
    }));
    await api.post(`/bill/${billId}/add-menu-items`, payload);
  },
  async getBillDetails(billId: number): Promise<BillDetailsResponse> {
    const response = await api.get<BillDetailsResponse>(`/bill/${billId}`);
    return response.data;
  },

  async addMenuItems(billId: number, menuItems: MenuItem[]): Promise<void> {
    if (menuItems == null || !menuItems.length) {
      throw new Error("No menu items provided.");
    }

    const bill = await this.getBill(billId);
    if (!bill) {
      throw new Error(`Bill with ID ${billId} not found.`);
    }

    await api.post(`/bill/${billId}/add-menu-items`, menuItems);
  },

  updateBillStatus: async (billId: number, status: 'pending' | 'settled'): Promise<void> => {
    await api.patch(`/bill/${billId}/status`, { status });
  },

  updateBillPaidBy: async (billId: number, paidBy: string): Promise<void> => {
    await api.patch(`/bill/${billId}/paid-by`, { paidBy });
  },

  updateBill: async (billId: number, updates: Partial<Bill>): Promise<void> => {
    await api.patch(`/bill/${billId}`, updates);
  },

  deleteBill: async (billId: number): Promise<void> => {
    await api.delete(`/bill/${billId}`);
  },

  async updateBillName(billId: number, name: string): Promise<void> {
    await api.patch(`/bill/${billId}/updata-name`, name);
  },

  async updateBillDate(billId: number, date: Date): Promise<void> {
    await api.patch(`/bill/${billId}/updata-date`, date);
  }
}; 